doc/arch/adr-004-geoprocessing-api.md

```diff
@@ -0,0 +1,56 @@
+
+MMW/BigCZ already includes and makes limited use of the Django REST Framework (DRF), which supports convenience methods around creating HTTP APIs.
+
+#### API Objectives
+* Support execution of all current and future "Analyze" sub tasks (Land Use, Soil, etc)
+* Support execution of Rapid Watershed Delineation
+* Model execution functionality will not be exposed
+* Available only for registered users
+* Usage will be recorded
+* Long term goal would be to charge for access
+
+#### Considerations
+* Geoprocessing routines are asynchronous, requiring submission and polling to determine when results are available.
+* Geoprocessing is a resource intensive operation.
+* Geoprocessing endpoints must remain open for anonymous users using the web platform.
+* Azavea maintains the only client of the current API, therefore we have had the freedom to make changes at will.  When an API is released, we'll have to treat releases of the platform carefully and possibly retain backwards compatibility.
+
+Based on the above, an API solution should adhere to the following implementation details:
+##### 1. Have a mechanism for authenticating a user
+Two options seem to be potentially appropriate: an API key associated with a user account, or an expiring token generated based on client supplied credentials.  API keys are easier to use, but inherently less secure since they can be reused by multiple parties and have to be included in the request. They can be revoked, allowing a failsafe against long term unauthorized use. Expiring tokens require the user to securely manage credentials, which can be difficult, especially when the desire is to use the API from frontend browser code.
+##### 2. Have at least rudimentary rate limiting capabilities
+Unrestricted job submission, even if used modestly, could have a considerable negative impact on the web platform performance.  If users anticipate using the API for batch processing, they should be restricted in how often they can submit.
+##### 3. Not result in duplicated code
+The web platform needs to have access to unauthenticated endpoints.  This could theoretically be accomplished through domain white listing within the DRF endpoints or the existing endpoints.  Regardless, the implementation should be generalized in one specific place that the API and platform utilize.
+##### 4. Enforce size and shape validation
+The front end code which submits analysis jobs is responsible for the majority of shape validation.  This will need to be included in the API, as well as for additional potential conflicts like coordinate systems and being contained within the CONUS.
```

Comment:
Do you mean ingest in a way that the app is currently not doing?  I'm picturing a geojson payload, and we do checks on area size, CONUS and self intersecting.  We may want to guard against extremely large payload sizes, which we could probably do via nginx.  Did you have something more in mind?